test/extensions/filters/http/ext_proc/ext_proc_grpc_fuzz.cc

```diff
@@ -0,0 +1,303 @@
+#include <vector>
+
+#include "envoy/config/core/v3/base.pb.h"
+#include "envoy/extensions/filters/http/ext_proc/v3alpha/ext_proc.pb.h"
+#include "envoy/service/ext_proc/v3alpha/external_processor.pb.h"
+#include "envoy/type/v3/http_status.pb.h"
+
+#include "source/common/network/address_impl.h"
+
+#include "test/common/http/common.h"
+#include "test/extensions/filters/http/ext_proc/ext_proc_grpc_fuzz_helper.h"
+#include "test/extensions/filters/http/ext_proc/test_processor.h"
+#include "test/fuzz/fuzz_runner.h"
+#include "test/integration/http_integration.h"
+#include "test/test_common/utility.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace HttpFilters {
+namespace ExternalProcessing {
+
+using envoy::config::core::v3::HeaderValue;
+using envoy::config::core::v3::HeaderValueOption;
+using envoy::extensions::filters::http::ext_proc::v3alpha::ProcessingMode;
+using envoy::service::ext_proc::v3alpha::BodyResponse;
+using envoy::service::ext_proc::v3alpha::CommonResponse;
+using envoy::service::ext_proc::v3alpha::HeadersResponse;
+using envoy::service::ext_proc::v3alpha::HttpBody;
+using envoy::service::ext_proc::v3alpha::HttpHeaders;
+using envoy::service::ext_proc::v3alpha::HttpTrailers;
+using envoy::service::ext_proc::v3alpha::ImmediateResponse;
+using envoy::service::ext_proc::v3alpha::ProcessingRequest;
+using envoy::service::ext_proc::v3alpha::ProcessingResponse;
+using envoy::service::ext_proc::v3alpha::TrailersResponse;
+using envoy::type::v3::StatusCode;
+
+using Http::LowerCaseString;
+
+// The buffer size for the listeners
+static const uint32_t BufferSize = 100000;
+
+// These tests exercise the ext_proc filter through Envoy's integration test
+// environment by configuring an instance of the Envoy server and driving it
+// through the mock network stack.
+
+class ExtProcIntegrationFuzz : public HttpIntegrationTest,
+                               public Grpc::BaseGrpcClientIntegrationParamTest {
+public:
+  ExtProcIntegrationFuzz(Network::Address::IpVersion ip_version, Grpc::ClientType client_type)
+      : HttpIntegrationTest(Http::CodecType::HTTP2, ip_version) {
+    ip_version_ = ip_version;
+    client_type_ = client_type;
+  }
+
+  void TearDown() {
+    cleanupUpstreamAndDownstream();
+    test_processor_.shutdown();
+  }
+
+  Network::Address::IpVersion ipVersion() const override { return ip_version_; }
+  Grpc::ClientType clientType() const override { return client_type_; }
+
+  void initializeFuzzer(bool autonomous_upstream) {
+    autonomous_upstream_ = autonomous_upstream;
+    autonomous_allow_incomplete_streams_ = true;
+    initializeConfig();
+    HttpIntegrationTest::initialize();
+  }
+
+  void initializeConfig() {
```

Comment:
My machine was getting around 3 exec/sec on a good day. I think one of the biggest things holding it back currently is when a mutation of the stream causes request/response to never complete. Off the top of my head one reason a connection will hang is if the external processor changes the value of end_stream from true to false.

Reply:
:< I don't even think persistent execution makes sense here given the randomized responses.

Reply:
Sorry, I don't think I understand. Could you explain what you mean by persistent execution?

Reply:
I just remembered, my inline 8-bit counters is really high for local testing. I was only able to get it down to about 780k counters from over 1M. I saw in the fuzzing report they were able to get it down to 300k or so. That could result in a major speed boost

Reply:
Persistent execution is where we keep the Envoy server threads around between executions of the fuzzer. This hurts determinism, but avoids the costly initialization/teardown of an Envoy server. I think it would be worth noting as something to consider later, similarly, please track the counters optimization.

Reply:
Note about persistent execution and counters were added to the TODO list